### ISSUE TITLE
Binary install of Ansible

### ISSUE THREAD

**Issue by kvz** [kvz/frey#51]
I just found this via that HN thread on simp_le: http://www.pyinstaller.org/ and think it might be great to distribute Frey's python components. Currently I have the problem (not sure if I mentioned already) that we changed nothing about Frey, but for auto-deployment of the uppy.io server, Travis is pulling in a newer python lib required by ansible through fuzzy version matching, and now [it is broken](https://travis-ci.org/transloadit/uppy-server/builds/168638885#L425) where it [worked before](https://travis-ci.org/transloadit/uppy-server/builds/167676412). Seems if we could build binaries of ansible and put those on S3 with pinned versions, the Frey experience might become more robust and quicker to install - thoughts?

Seems someone tried this before and ran into roadblocks: https://groups.google.com/forum/#!msg/ansible-project/VOViIDNG1Wc/wCPgOcN_ESEJ - but perhaps there is a workaround in the proposed ` --hidden-import=callbackModule.py` arguments

@tersmitten Would it be better if I created GitHub issues for these, so we can have threaded discussions in an async fashion? Or is this channel fine for you (edited)

**Comment by tersmitten** [comment]
Maybe these are useful: 

* https://pantsbuild.github.io/python-readme.html
* https://github.com/jordansissel/fpm

**Comment by kvz** [comment]
We just had somebody run into this https://github.com/pyca/cryptography/issues/2350#issuecomment-169661673 - the binary thing seems quite important to have

**Comment by kvz** [comment]
Someone else running into 

```
cc -fno-strict-aliasing -fno-common -dynamic -arch i386 -arch x86_64 -g -Os -pipe -fno-common -fno-strict-aliasing -fwrapv -DENABLE_DTRACE -DMACOSX -DNDEBUG -Wall -Wstrict-prototypes -Wshorten-64-to-32 -DNDEBUG -g -fwrapv -Os -Wall -Wstrict-prototypes -DENABLE_DTRACE -arch i386 -arch x86_64 -pipe -DUSE__THREAD -I/usr/include/ffi -I/usr/include/libffi -I/System/Library/Frameworks/Python.framework/Versions/2.7/include/python2.7 -c c/_cffi_backend.c -o build/temp.macosx-10.12-intel-2.7/c/_cffi_backend.o
    c/_cffi_backend.c:15:10: fatal error: 'ffi.h' file not found
    #include <ffi.h>
             ^
    1 error generated.
    error: command 'cc' failed with exit status 1
    
    ----------------------------------------
Command "/usr/bin/python -u -c "import setuptools, tokenize;__file__='/private/var/folders/j7/7zwh5brd6jvgh11t1v41nw4w0000gp/T/pip-build-p7b9na/cffi/setup.py';exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" install --record /var/folders/j7/7zwh5brd6jvgh11t1v41nw4w0000gp/T/pip-r9j2Sa-record/install-record.txt --single-version-externally-managed --root /Users/ifedapo/.frey/tools/ansible/2.0.2.0 --compile --prefix=pip" failed with error code 1 in /private/var/folders/j7/7zwh5brd6jvgh11t1v41nw4w0000gp/T/pip-build-p7b9na/cffi/

--> Exiting with error: Failed to install 'ansible'. Error: Script 'bash -o pipefail -o errexit -o nounset -c mkdir -p /Users/ifedapo/.frey/tools/ansible/2.0.2.0 && pip install --install-option='--prefix=pip' --ignore-installed --force-reinstall --root '/Users/ifedapo/.frey/tools/ansible/2.0.2.0' --upgrade --disable-pip-version-check ansible==2.0.2.0' exited with code: '1'
Cleaning up '/Users/ifedapo/projects/uppy-server/infra/ssh/frey-uppy-server.pem' after process exit with code '1' 
```

http://stackoverflow.com/questions/22875270/error-installing-bcrypt-with-pip-on-os-x-cant-find-ffi-h-libffi-is-installed

**Comment by kvz** [comment]
I wonder if `virtualenv` could already help solve this to an extend, maybe @ifedapoolarewaju has ideas on this?

**Comment by tersmitten** [comment]
I always run Ansible in `virtualenv`, but then you have `pip` and `virtualenv` as a dependencies

**Comment by ifedapoolarewaju** [comment]
virtualenv only helps solve issues with python related dependencies and path issues. Virtualenv would not solve a problem like [this](https://github.com/freyproject/frey/issues/51#issuecomment-259699396) which actually required installation of a different OS library.

**Comment by tersmitten** [comment]
These are [the dependencies](https://github.com/Oefenweb/ansible-ansible/blob/master/vars/main.yml) we install to install Ansible

**Comment by ifedapoolarewaju** [comment]
Yes these dependencies seem like they may be required for some python libraries that have `C extensions` not withstanding, these dependencies are os libraries. So if a pip installation fails due to a `C Library/OS Library` irregularity, I'm afraid a virtualenv cannot help much in this case because the virtualenv does not control/manage such global libraries. It only manages the python libraries that may need them.

**Comment by ifedapoolarewaju** [comment]
So I ran some experiment combining various installation conditions for ansible and I thought to share my findings.

### 1
*pipVersion*: 7.x

*CMD*: `pip install --no-cache-dir --install-option='--prefix=pip' --ignore-installed --force-reinstall --root '/Users/ifedapo/.ansible-test/2.1.3' --upgrade --disable-pip-version-check ansible==2.1.3`

*Warning*: ` Disabling all use of wheels due to the use of --build-options / --global-options / --install-options.`

*Error*: 
```
build/temp.macosx-10.12-intel-2.7/_openssl.c:433:10: fatal error: 'openssl/opensslv.h' file not found
    #include <openssl/opensslv.h>
             ^
    1 error generated.
    error: command 'cc' failed with exit status 1
    
Command "/Users/ifedapo/.virtualenvs/test-ansible/bin/python -c "import setuptools, tokenize;__file__='/private/var/folders/j7/7zwh5brd6jvgh11t1v41nw4w0000gp/T/pip-build-KN0R6f/cryptography/setup.py';exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" install --record /var/folders/j7/7zwh5brd6jvgh11t1v41nw4w0000gp/T/pip-uf2YwC-record/install-record.txt --single-version-externally-managed --root /Users/ifedapo/.ansible-test/2.1.3 --compile --install-headers /Users/ifedapo/.virtualenvs/test-ansible/include/site/python2.7/cryptography --prefix=pip" failed with error code 1 in /private/var/folders/j7/7zwh5brd6jvgh11t1v41nw4w0000gp/T/pip-build-KN0R6f/cryptography
```

### 2

*pipVersion*: 7.x

*CMD*: `pip install --no-cache-dir ansible==2.1.3`

*Warning*: none

*Error*: 
```
 build/temp.macosx-10.12-intel-2.7/_openssl.c:433:10: fatal error: 'openssl/opensslv.h' file not found
    #include <openssl/opensslv.h>
             ^
    1 error generated.
    error: command 'cc' failed with exit status 1
    
    ----------------------------------------
Command "/Users/ifedapo/.virtualenvs/test-ansible/bin/python -c "import setuptools, tokenize;__file__='/private/var/folders/j7/7zwh5brd6jvgh11t1v41nw4w0000gp/T/pip-build-on9wTa/cryptography/setup.py';exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" install --record /var/folders/j7/7zwh5brd6jvgh11t1v41nw4w0000gp/T/pip-E_bgIW-record/install-record.txt --single-version-externally-managed --compile --install-headers /Users/ifedapo/.virtualenvs/test-ansible/include/site/python2.7/cryptography" failed with error code 1 in /private/var/folders/j7/7zwh5brd6jvgh11t1v41nw4w0000gp/T/pip-build-on9wTa/cryptography
```

### 3

*pipVersion*: 9.x

*CMD*: `pip install --no-cache-dir ansible==2.1.3`

*Warning*: none

*Error*: none

### 4

*pipVersion*: 9.x

*CMD*: `pip install --no-cache-dir --install-option='--prefix=pip' --ignore-installed --force-reinstall --root '/Users/ifedapo/.ansible-test/2.1.3' --upgrade --disable-pip-version-check ansible==2.1.3`

*Warning*: `/Users/ifedapo/.virtualenvs/test-ansible/lib/python2.7/site-packages/pip/commands/install.py:194: UserWarning: Disabling all use of wheels due to the use of --build-options / --global-options / --install-options`

*Error*: 
```
#include <openssl/opensslv.h>
             ^
    1 error generated.
    error: command 'cc' failed with exit status 1
    
    ----------------------------------------
Command "/Users/ifedapo/.virtualenvs/test-ansible/bin/python -u -c "import setuptools, tokenize;__file__='/private/var/folders/j7/7zwh5brd6jvgh11t1v41nw4w0000gp/T/pip-build-kUVcV5/cryptography/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /var/folders/j7/7zwh5brd6jvgh11t1v41nw4w0000gp/T/pip-ez49XW-record/install-record.txt --single-version-externally-managed --root /Users/ifedapo/.ansible-test/2.1.3 --compile --install-headers /Users/ifedapo/.virtualenvs/test-ansible/include/site/python2.7/cryptography --prefix=pip" failed with error code 1 in /private/var/folders/j7/7zwh5brd6jvgh11t1v41nw4w0000gp/T/pip-build-kUVcV5/cryptography/
```

### 5

*pipVersion*: 9.x

*CMD*: `pip install --no-cache-dir --ignore-installed --force-reinstall --root '/Users/ifedapo/.ansible-test/2.1.3' --upgrade --disable-pip-version-check ansible==2.1.3`

*Warning*: none

*Error*: none

### 6

*pipVersion*: 7.x

*CMD*: `pip install --no-cache-dir --ignore-installed --force-reinstall --root '/Users/ifedapo/.ansible-test/2.1.3' --upgrade --disable-pip-version-check ansible==2.1.3`

*Warning*: none

*Error*: 
```
build/temp.macosx-10.12-intel-2.7/_openssl.c:433:10: fatal error: 'openssl/opensslv.h' file not found
    #include <openssl/opensslv.h>
             ^
    1 error generated.
    error: command 'cc' failed with exit status 1
    
    ----------------------------------------
Command "/Users/ifedapo/.virtualenvs/test-ansible/bin/python -c "import setuptools, tokenize;__file__='/private/var/folders/j7/7zwh5brd6jvgh11t1v41nw4w0000gp/T/pip-build-XkL1V8/cryptography/setup.py';exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" install --record /var/folders/j7/7zwh5brd6jvgh11t1v41nw4w0000gp/T/pip-WaCcF0-record/install-record.txt --single-version-externally-managed --root /Users/ifedapo/.ansible-test/2.1.3 --compile --install-headers /Users/ifedapo/.virtualenvs/test-ansible/include/site/python2.7/cryptography" failed with error code 1 in /private/var/folders/j7/7zwh5brd6jvgh11t1v41nw4w0000gp/T/pip-build-XkL1V8/cryptography
```

## TL;DR

There seem to be 2 main issues. The pip version and the `--install-option='--prefix=pip'` flag.

I am not sure why we need a lower version of pip, maybe it is a special requirement for this project? But as it seems, it appears to be less reliable when handling installations. The installation always fails with this version. This is the case where we then have to use [this](https://github.com/pyca/cryptography/issues/2350#issuecomment-169661673) as a solution where as this wouldn't be required for higher version `>=8.x`

As per the flag `--install-option='--prefix=pip'` I get the feeling that this is where `virtualenv` may come in handy. It seems this flag is used because we have a custom pip directory. But the thing is this flag disables `whl` format of python library installations and uses `.egg` instead. whl is more recent and effecient see [this](https://packaging.python.org/wheel_egg/). And apparently it is more reliable because when the `--install-option` flag is disabled and wheel is used for the installations no failures occur. As opposed to when `.egg` is used. And if `--install-option` is needed because of a custom `pip directory` then maybe we can use `virtualenv` instead, which solves all the problems.


I hope this information is useful. And in the event that this was already known earlier, then I guess this has been quite some use of time 😃

**Comment by kvz** [comment]
Wow, thanks a lot @ifedapoolarewaju! Seems that a newer pip and using virtual env will already solve a bunch of problems. Pip upgrade is a very straightforward in `src/Deps.js`, but I'm unsure how virtualenv would look. Mischa explained this means adding another dependency (virtualenv) to the project, installing that globally via pip (now we're touching a global install? 😢 or can this be done local? or is it available always as part of python/pip?)

After we know how to get the virtualenv dep installed across platforms, the question is how the commandline for this would work. Just prefix the usual commands with `virtualenv`? Sorry, I'm quite the python noob.

**Comment by ifedapoolarewaju** [comment]
@kvz So I did a little more research and it seems this flag `--prefix` got released with pip8 https://github.com/pypa/pip/pull/3252 . This has a general support as opposed to `--install-option='--prefix=pip'`.

My new command looks like this

`pip install --no-cache-dir --prefix=pip --ignore-installed --force-reinstall --root '/Users/ifedapo/.ansible-test/2.1.3' --upgrade --disable-pip-version-check ansible==2.1.3`

And it behaves as desired. I mean running
```ls /Users/ifedapo/.ansible-test/2.1.3/pip -->
bin	lib```

So the proposed solution wouldn't need `virtualenv` it would only require to upgrade pip and replace the `--install-option='--prefix=pip'` flag with just `--prefix=pip`

**Comment by kvz** [comment]
@ifedapoolarewaju that's great Ife thanks so much for diving into this! ❤️ I'm currently thinking of [adding OSX](https://docs.travis-ci.com/user/osx-ci-environment/#OS-X-Version) to the build matrix of Travis CI, executing only the `prepare` step there, where all the installing of dependencies take place - seeing it fail - then adding your proposed method to hopefully see it pass. Will work on that soonish

**Comment by kvz** [comment]
Alright its' failing. Pip is installed, and then pip isn't found upon verification: https://travis-ci.org/freyproject/frey/jobs/175462479#L1187

Pip 9 i installed though.. This needs some more debugging I fear

**Comment by kvz** [comment]
Ignore that last comment, i linked to a trusty build, that i also added. I'll investigate some more later on

**Comment by kvz** [comment]
Added your ideas @ifedapoolarewaju and the good news is that it's building on two Ubuntu versions:

![](https://dl.dropboxusercontent.com/s/73y6bstdc6g5fzx/2016-11-16%20at%2016.24.png)

Also added an OSX build, which will be more error prone, but the jury is still out on that one: https://travis-ci.org/freyproject/frey/builds/176400663

**Comment by bborysenko** [comment]
FYI: ansible/ansible#18502

**Comment by ifedapoolarewaju** [comment]
@kvz it seems the build failed for a different reason?

**Comment by kvz** [comment]
@bborysenko Thanks so much for that! I'm afraid they're going to make you jump through hoops though with that whole "it's gotta be a  https://github.com/ansible/proposals/blob/master/proposals_process_proposal.md" :/ 

@ifedapoolarewaju A later build passed! Here you can see it in all its glory https://travis-ci.org/freyproject/frey/builds/176493942 😄

**Comment by ifedapoolarewaju** [comment]
yippee! ✌🏽

**Comment by kvz** [comment]
https://github.com/pypa/pipfile/blob/master/README.rst

**Comment by kvz** [comment]
http://conda.pydata.org/docs/install/quick.html
